singularity-eos/eos/modifiers/zsplit_eos.hpp

```diff
@@ -0,0 +1,274 @@
+    const bool press_input = !press_output;
+
+    const bool cv_output = (output & thermalqs::specific_heat);
+    const bool bmod_output = (output & thermalqs::bulk_modulus);
+
+    if (sie_input) energy *= iscale;
+    if (press_input) press *= iscale;
+
+    t_.FillEos(rho, temp, energy, press, cv, output, lambda);
+
+    // do it to undo if it was input. do it because you need to if
+    // output
+    energy *= scale;
+    press *= scale;
+    // These aren't input
+    if (cv_output) cv *= scale;
+    if (bmod_output) bmod *= scale;
+  }
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION void
+  ValuesAtReferenceState(Real &rho, Real &temp, Real &sie, Real &press, Real &cv,
+                         Real &bmod, Real &dpde, Real &dvdt,
+                         Indexer_t &&lambda = static_cast<Real *>(nullptr)) const {
+    const Real scale = GetScale_(lambda);
+    t_.ValuesAtReferenceState(rho, temp, sie, press, cv, bmod, dpde, dvdt);
+    sie *= scale;
+    press *= scale;
+    cv *= scale;
+    bmod *= scale;
+    // dvdt, dpde unchanged
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  int nlambda() const noexcept { return NL + t_.nlambda(); }
+  static constexpr unsigned long PreferredInput() { return T::PreferredInput(); }
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    return T::scratch_size(method, nelements);
+  }
+  static inline unsigned long max_scratch_size(unsigned int nelements) {
+    return T::max_scratch_size(nelements);
+  }
+  PORTABLE_FUNCTION void PrintParams() const { t_.PrintParams(); }
+
+  PORTABLE_FORCEINLINE_FUNCTION Real MinimumDensity() const {
+    return t_.MinimumDensity();
+  }
+  PORTABLE_FORCEINLINE_FUNCTION Real MinimumTemperature() const {
+    return t_.MinimumTemperature();
+  }
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION Real MeanAtomicMassFromDensityTemperature(
+      const Real rho, const Real temperature,
+      Indexer_t &&lambda = static_cast<Real *>(nullptr)) const {
+    return t_.MeanAtomicMassFromDensityTemperature(rho, temperature, lambda);
+  }
+  template <typename Indexer_t = Real *>
+  PORTABLE_INLINE_FUNCTION Real MeanAtomicNumberFromDensityTemperature(
+      const Real rho, const Real temperature,
+      Indexer_t &&lambda = static_cast<Real *>(nullptr)) const {
+    return t_.MeanAtomicNumberFromDensityTemperature(rho, temperature, lambda);
+  }
+
+  SG_ADD_MODIFIER_METHODS(T, t_);
+  SG_ADD_MODIFIER_MEAN_METHODS(t_)
+
+ private:
+  static constexpr const int NL = 1;
+  template <typename Indexer_t = Real *>
+  PORTABLE_FORCEINLINE_FUNCTION Real GetIonizationState_(Indexer_t &&lambda) const {
+    return std::max(0.0, lambda[t_.nlambda()]);
+  }
+  // TODO(JMM): Runtime?
+  template <typename Indexer_t = Real *>
+  PORTABLE_FORCEINLINE_FUNCTION Real GetScale_(Indexer_t &&lambda) const {
+    Real Z = GetIonizationState_(lambda);
+    if constexpr (ztype == ZSplitComponent::Electrons) {
```

Comment:
Notice the if constexpr here, which makes this a compile-time switch rather than runtime. We could make it runtime, but I chose not to. The trade-off here is that run-time would make the variant smaller. But compile time increases performance.